course-material/intro-to-sql/aggregate-functions-in-sql.md

```diff
@@ -1,14 +1,287 @@
+extract_to_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/00222/bank.zip', bank)
+  ```
+After running this code, you should have `bank_cleaned.csv` in the current directory. 
+
+### Load Engine
+We now load in our SQL extension that allows us to execute SQL queries in Jupyter Notebooks. 
+
+<b>Note</b> Ensure you restart any previous notebook that has the same database name as the one initialized below.
+
+```{code-cell} ipython3
+# Loading in SQL extension
+%reload_ext sql
+# Initiating a DuckDB database named 'bank.duck.db' to run our SQL queries on
+%sql duckdb:///bank.duck.db
+```
+
+Let's now return to our initial dataset of bank marketing records. 
+
+### Queries
+
+#### Creating Table
+
+Let's start off with loading our `bank_cleaned.csv` file from our local directory to our newly created DuckDB database. Here we `CREATE OR REPLACE TABLE` in DuckDB called 'bank' `FROM` our `bank_cleaned.csv` file. The `read_csv_auto` is a function that helps SQL understand our local .csv file for creation into our database.
+
+```{code-cell} ipython3
+%%sql
+CREATE OR REPLACE TABLE bank AS
+FROM read_csv_auto('bank_cleaned.csv', header=True, sep=',')
+```
+
+#### Count and Distinct Count
+`COUNT()` allows users to query the count of records in a given query. A common argument to this function is `*`. `COUNT(*)` tells SQL to return the total number of rows from our query, including NULL values. This can be combined with queries that filter for certain conditions.
+
+A common function to be used with `COUNT()` is `DISTINCT()`. For example, selecting `COUNT(DISTINCT(job))` would return the count of distinct values under the "job" column.
+
+```{code-cell} ipython3
+%%sql 
+SELECT COUNT(DISTINCT(job))
+FROM bank 
+```
+
+The output of the above query can be an eyesore. Instead of getting SQL's default column name when running these aggregation function, users can utilize the `AS` clause. The text following `AS` changes the output of the `SELECT` clause to that text. Our next example with demonstrate this.
+
+Here's an example with `COUNT()` that counts the number of rows of our query `WHERE` we filter for "balance" greather than or equal to 500 `AND` where "martial" equals married. We also use the `AS` clause to change the column of our query to "Count". 
+
+```{code-cell} ipython3
+%%sql 
+SELECT COUNT(*) AS Count
+FROM bank 
+WHERE balance > 500 AND martial = 'married'
+```
+
+`COUNT()` can also have the arguments of a single row, such as `COUNT(job)`. `COUNT(job)` would count the number of rows of just the "job" column. If "job" were to have any NULL values in its query, those NULL values would be subtracted from the total row counts of the query. 
+
+#### Average
+
+`AVG()` allows users to take the average of columns. This clause can also be used with filtering. An example of finding the average balance of unemployed observations is as follows:
+
+```{code-cell} ipython3
+%%sql 
+SELECT AVG(balance) AS 'Average Unemployed Balance'
```

Comment:
Be careful with the use of spaces in a variable name. Good practice to use camel case ("average_unemployed_balance")

I have had DuckDB yell at me for including spaces instead of _ in my column names

You can use special characters if you really want to preserve the empty space, but it makes the data inconsistent